fawltydeps/packages.py

```diff
@@ -261,15 +261,24 @@ def packages(self) -> Dict[str, Package]:
         # that map to zero import names.
         context = DistributionFinder.Context(path=paths)  # type: ignore
         for dist in MetadataPathFinder().find_distributions(context):  # type: ignore
+            normalized_name = Package.normalize_name(dist.name)
             parent_dir = dist.locate_file("")
+            if normalized_name in ret:
+                # We already found another instance of this package earlier in
+                # the given paths. Assume that the earlier package is what
+                # Python's import machinery will choose, and that this later
```

Comment:
Why do we assume so?

Reply:
Principle of least surprise: This code is about finding modules/packages in `sys.path` (or a similar list of package directories passed to `DistributionFinder.Context()`). When you do `import foobar` in Python, Python itself traverses `sys.path` until it find the _first_ matching `foobar` module/package, and then imports (only) that one. If there is another `foobar` further back in `sys.path` it is never found/used by Python itself.

Before this change our code did the exact opposite: The latter matches would overwrite the earlier matches in the `ret` dict, and you would end up with `Package` objects that reflected different versions/locations than what Python itself would actually use at runtime. Although we _could_ merge the `Package` objects from multiple sources, that seems like the wrong thing to do here, as Python itself would never see/use anything but the first match.

Note that we here talk about resolving packages/modules in a _single_ Python environment, which in the case of `sys.path` is a _single_ Python environment that consists of _multiple_ package directories (with a defined priority/order).

Later in this PR, we introduce the concept of multiple Python environments, and then our rules are different:

If you pass `--pyenv dirA --pyenv dirB`, we assume that _both_ environments are somehow relevant to your project: It may be that you run your project in _either_ `dirA` _or_ `dirB`, or it might be that `dirA` represents your prod environment and `dirB` represents your test environment. No matter, we treat `dirA` and `dirB` as _two separate Python environments of equal importance_, and we therefore need to merge their `Package` objects accordingly. (But that is _not_ what is happening in _these_ lines of code, rather that is handled in a later commit: with the refactoring to `_from_one_env()` in https://github.com/tweag/FawltyDeps/pull/313/commits/ec726e69c9e62c45c726eae04e5539b769f58b35.)

Reply:
Thanks for the explanation :+1:  I did not know how Python does it internally, if it takes first or the last.